controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go

```diff
@@ -153,6 +153,7 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
 		{spec, "machineTemplate", "metadata", "*"},
 		{spec, "machineTemplate", "infrastructureRef", "apiVersion"},
 		{spec, "machineTemplate", "infrastructureRef", "name"},
+		{spec, "machineTemplate", "infrastructureRef", "kind"},
```

Comment:
hi @sbueringer, we briefly touched on the idea of adding infrastructureRef:kind to allow paths. It was blocked because infrastructureRef:kind change usually indicates an infra provider change. In our cases, there was a kind change but no infra provider change. It is OK to add it to allow paths?  I was also thinking about adding some logics to check if the kinds are belong to the same infra provider but that would need upstream needs to beware of downstream kinds which may not be ideal.

any insights would be greatly appreciated. thanks!

Reply:
I would prefer to not allow this change by default.
What we can do is to make this possible via dedicated annotation on the KCP object, something like "unsafe.kcp.cluster.x-k8s.io/disable-update-infrastructureRef-kind-check"
@sbueringer @vincepri opinions?

Reply:
Good question.

I wonder how often folks are intentionally changing the kind in KCP. Even if they do that it either works or it fails on the first machine that KCP rolls out.

So I'm not sure if it's KCP's concern to figure out if different kinds are compatible. It's already allowed to change the apiGroup. 


So for me it's a tradeoff between stop caring in KCP about compatibility in infrastructure kinds vs. introducing a new API with that annotation to conditionally allow changes in the kind.


The latter option is definitely safer as it requires additional opt-in, but as mentioned we already allow apiGroup changes (if I see correctly) which makes even less sense (I know we do it for the version and it's just one field).


So I think overall I would slightly prefer to allow kind-changes as it doesn't seem to be much unsafer then apiGroup changes and I would prefer to avoid having too many addition "flags" like annotations.

But no objections either way if folks have a different preference.

Reply:
@fabriziopandini @vincepri do you have any preference?

Reply:
I kind of agree with @sbueringer argument that allowing change of kind is no more dangerous than allowing change of apiVersion (which we are required to do for handling provider upgrades).
So I'm fine with keeping the change as it is